chartsxblock/static/js/src/chartsxblock-studio.js

```diff
@@ -37,36 +99,117 @@ function BadChartError(message) {
-    return [row_count, Math.min.apply(null, row_sizes)]
+    if(Math.min.apply(null, row_sizes) != Math.max.apply(null, row_sizes))
+        throw new BadChartError("All chart rows are not equal length.");
+    return [row_count, Math.min.apply(null, row_sizes)];
 }
-*/
 
 function ValidateData(inputString) {
-    //TODO
     try {
         parsedString = JSON.parse(inputString);
     } catch (e) {
         throw e;
     }
+    chartDimensions = size(parsedString);
+    if (chartDimensions[0] < 2) throw new BadChartError("The chart requires more than one row.")
     return inputString;
 }
 
-function ValidateType(inputString) {
-    //TODO
+function ValidateType(inputString, chartTypes) {
+    if(chartTypes.indexOf(inputString) == -1 ) throw new BadChartError("Using an unsupported chart type.");
     return inputString;
 }
 
 function ValidateName(inputString) {
-    //TODO
+    if(inputString.length < 1) throw new BadChartError("Chart name is empty.");
     return inputString;
 }
+
+function ChartTable(targetDiv, inputJSON){
+    this.data = JSON.parse(inputJSON);
+    this.rows = this.data.length;
+    this.columns = this.data[0].length;
+    this.div = targetDiv;
+
+    this.draw = function() {
+    // Draws the table in a div
+        for (i = 0; i < this.data.length; i++) {
+            var row = $("<tr>");
+            for (j = 0; j < this.data[i].length; j++) {
+                var field = $("<td>" + this.data[i][j] + "</td>");
+                row.append(field);
+            }
+            this.div.find("tbody").append(row);
+        }
+    }
+
+    this.update = function() {
+    // Updates the table from its div element
+        var dataArray = [];
+        var parent = this;
+        this.div.find("tbody").children("tr").each(function() {
+            var row = [];
+            $(this).children("td").each(function() {
+                row.push(parent.parse($(this).text()));
+            });
+            dataArray.push(row);
+        });
+        this.data = dataArray;
+    }
+
+    this.parse = function(val) {
+        if (val == parseInt(val)) {
+            return parseInt(val);
+        }
+        else if (val == parseFloat(val)) {
+            return parseFloat(val);
+        }
+        else {
+            return val;
+        }
+    }
+
+    this.json = function(skipUpdate) {
+    // If called without parameters - updates then sends a JSON
+        if (typeof(toUpdate)==='undefined') this.update();
+        return(JSON.stringify(this.data));
+    }
+
+    this.addRow = function() {
+        var rowSelector = $('<tr>');
+        this.div.find("tbody").append(rowSelector);
+        for (i = 0; i < this.columns; i++) {
+            rowSelector.append($('<td>').makeEditable());
+        }
+        this.rows++;
+    }
+
+    this.removeRow = function() {
+        if(this.rows <2) throw {name: "Dimensions Error", message: "You cannot have 0 rows."};
+        this.div.find("tbody").find("> tr:last").remove();
```

Comment:
Spojio i ovdje, i na ostalim mjestima. Ostalo radi navike i čitljivosti. (iako što se performansi tiče, ne bi trebalo biti neke razlike)